activerecord/lib/active_record/relation/predicate_builder/polymorphic_array_handler.rb

```diff
@@ -0,0 +1,57 @@
+module ActiveRecord
+  class PredicateBuilder
+    class PolymorphicArrayHandler # :nodoc:
+      def initialize(predicate_builder)
+        @predicate_builder = predicate_builder
+      end
+
+      def call(attribute, value)
+        table = value.associated_table
+        queries = value.type_to_ids_mapping.map do |type, ids|
+          { table.association_foreign_type.to_s => type, table.association_foreign_key.to_s => ids }
+        end
+
+        predicates = queries.map { |query| predicate_builder.build_from_hash(query) }
+
+        if predicates.size > 1
+          type_and_ids_predicates = predicates.map { |type_predicate, id_predicate| Arel::Nodes::Grouping.new(type_predicate.and(id_predicate)) }
+          type_and_ids_predicates.inject(&:or)
+        else
+          predicates.first
+        end
```

Comment:
On the first iteration, `result` is the array of predicates so it raises: `NoMethodError: undefined method `or' for #<Array:0x007f874b75b348>`.  I can't think of a good initial value to pass to reduce that would avoid this problem without adding an extra `or`.  

I could remove the `if predicates.size > 1` but when there is only one object passed, the resulting sql gets an extra set of `()` and it breaks a test.  Happy to remove the conditional and update the test to assert against the results instead of the generated sql as it sounds like that's preferred now.  LMK.